corehq/apps/hqadmin/tasks.py

```diff
@@ -180,3 +181,25 @@ def track_es_doc_counts():
                     }
                     metrics_gauge('elasticsearch.shards.docs.count', i['docs']['count'], tags)
                     metrics_gauge('elasticsearch.shards.docs.deleted', i['docs']['deleted'], tags)
+
+
+@periodic_task(queue='background_queue', run_every=crontab(minute="0", hour="0"))
+def track_pg_limits():
+    for db in settings.DATABASES:
+        with connections[db].cursor() as cursor:
+            query = """
+            select tab.relname, seq.relname, attlen
+              from pg_class seq
+              join pg_depend dep on seq.oid=dep.objid
+              join pg_class as tab on (dep.refobjid = tab.oid)
+              join pg_attribute on attrelid=tab.oid and attnum=refobjsubid
+              where seq.relkind='S'
+            """
+            cursor.execute(query)
+            results = cursor.fetchall()
+            for table, sequence, length in results:
+                # Only track integer columns, ignoring ones that are already bigint
+                if length == 4:
+                    cursor.execute(f'select last_value from {sequence}')
+                    current_value = cursor.fetchone()[0]
+                    metrics_gauge('postgres.sequence.current_value', current_value, {'table': table, 'database': db})
```

Comment:
Brilliant. So to summarize it looks like for each table with a sequence, you gt the table name, sequence name, and the length of the primary key column of the sequence, and then (only if the length is 4) you fetch the next value from the sequence and using that as the high water mark for that column. Makes sense, took me a little bit of staring at my screen to follow.

I'm curious—could you achieve the filtering by length by putting it in the original query and then not selecting length at all from the query? Could you embed the `select last_value from {sequence}` part in the query? Not even sure if either of those are better, so if neither of those spark anything for you, you can just ignore it.

Reply:
That sounds right to me. I think we could definitely add the length filtering to the main query. I do not think we could add the other part since the dynamic bit is in the `from` clause (like a table) rather than a set of rows we can join on.

Reply:
> and the length of the primary key column of the sequence
it is the length of the column that owns the sequence (in theory that doesn't have to be a primary key but I imagine that is the only place we use sequences)

Reply:
Nice I like how incorporating the length = 4 bit into the query made it a bit simpler.

And noted about the column that owns a seq, not necessarily only primary keys, thanks. It's true that if that case ever does happen we'd want to track that too, so that makes sense.